pkg/baseconn/conn.go

```diff
@@ -0,0 +1,162 @@
+// Copyright 2019 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package baseconn
+
+import (
+	"database/sql"
+
+	tcontext "github.com/pingcap/dm/pkg/context"
+	"github.com/pingcap/dm/pkg/log"
+	"github.com/pingcap/dm/pkg/retry"
+	"github.com/pingcap/dm/pkg/terror"
+
+	"go.uber.org/zap"
+)
+
+// BaseConn wraps a connection to DB
+type BaseConn struct {
+	DB *sql.DB
+
+	// for reset
+	DSN string
+
+	RetryStrategy retry.Strategy
+}
+
+// SQL is format sql job
+type SQL struct {
+	Query string
+	Args  []interface{}
+}
+
+// NewBaseConn builds BaseConn to connect real DB
+func NewBaseConn(dbDSN string, strategy retry.Strategy) (*BaseConn, error) {
+	db, err := sql.Open("mysql", dbDSN)
+	if err != nil {
+		return nil, terror.ErrDBDriverError.Delegate(err)
+	}
+	err = db.Ping()
+	if err != nil {
+		db.Close()
+		return nil, terror.ErrDBDriverError.Delegate(err)
+	}
+	if strategy == nil {
+		strategy = &retry.FiniteRetryStrategy{}
+	}
+	return &BaseConn{db, dbDSN, strategy}, nil
+}
+
+// SetRetryStrategy set retry strategy for baseConn
+func (conn *BaseConn) SetRetryStrategy(strategy retry.Strategy) error {
+	if conn == nil {
+		return terror.ErrDBUnExpect.Generate("database connection not valid")
+	}
+	conn.RetryStrategy = strategy
+	return nil
+}
+
+// ResetConn generates new *DB with new connection pool to take place old one
+func (conn *BaseConn) ResetConn(tctx *tcontext.Context) error {
+	if conn == nil {
+		return terror.ErrDBUnExpect.Generate("database connection not valid")
+	}
+	db, err := sql.Open("mysql", conn.DSN)
+	if err != nil {
+		return terror.ErrDBDriverError.Delegate(err)
+	}
+	err = db.Ping()
+	if err != nil {
+		db.Close()
+		return terror.ErrDBDriverError.Delegate(err)
+	}
+	if conn.DB != nil {
+		err := conn.DB.Close()
+		if err != nil {
+			tctx.L().Warn("reset connection", log.ShortError(err))
+		}
+	}
+	conn.DB = db
+	return nil
+}
+
+// QuerySQL defines query statement, and connect to real DB
+func (conn *BaseConn) QuerySQL(tctx *tcontext.Context, query string, args ...interface{}) (*sql.Rows, error) {
```

Comment:
ExecuteSQL has multiple queries and args, but QuerySQL has only one query and variable arguments, if we change `query` and `args` to `sql`, it may lost variable arguments(...args)

Reply:
if so, how about let all of them use `query` and `args` directly? Is `SQL` struct necessary?

Reply:
you are right, I removed `SQL`, it's seems better now